### ISSUE TITLE
Negative loss and unstable training

### ISSUE THREAD

**Issue by dragonlong** [nvidia-cosmos/cosmos-reason1#22]
Hi, 

I am trying to follow the RL training using the default dataset. However, at my first try, it seems that there are some **fundamental errors** with the implementation and training scripts, like the loss can be negative and it is very unstable, as attached below. Compared to the Cosmos-Reason1 paper results, I am wondering what went wrong here, or it is kind expected? `python tools/launch_all.py --config configs/cosmos-reason1/cosmos-reason1-7b-p-fsdp1-tp2-r-tp2-pp1-grpo.toml --launcher ./tools/dataset/cosmos_grpo.py`, and I use 4 A100 GPUs. The default dataset is *nvidia/Cosmos-Reason1-RL-Dataset* with *robovqa*.

Thanks

![Image](https://github.com/user-attachments/assets/a046a158-68de-4283-9a1e-53da8ca1414c)

**Comment by jcao-ai** [comment]
Loss computation of GRPO will be influenced by a few factors:
1. Loss type: `token-mean`, `seq-mean-token-sum`, `seq-mean-token-mean`;
2. Batch size vs `n_generation`.

## Loss type 

<img width="720" height="92" alt="Image" src="https://github.com/user-attachments/assets/a0862fbe-6e8f-4eb5-b74f-18cc2473c4f8" />


The likelihood ratios are all ones since only one iteration (fully on-policy), and the first part of the loss sums to zero. In this case, `avg_loss`  equals `beta * KL` , since KL is always positive, the `avg_loss` is always positive as well. 

<img width="719" height="93" alt="Image" src="https://github.com/user-attachments/assets/f3b10691-41ca-4b44-9b4c-00246ae07a2f" />

However, this conclusion does not apply when loss type is `seq-mean-token-sum` or `seq-mean-token-mean`, since loss will take sample length into consideration and loss could sum to non-zero values in that case.

## Batch size vs `n_generation`
If global batch size is smaller than `n_generation`, which means a group of samples is divided into multi-batches, will definitely lead to `sum(adv_i) != 0`.
